### PR TITLE
ep testing

### PR DESCRIPTION
**Pull request by Sidu28**
More EP testing, reconfigure the CI to work efficiently and not take 10 minutes lol.  It now takes 2

### CODE REVIEW

**Comment by ChaoticWalrus** [comment]
@Sidu28 Cool idea with the parallelization of running the tests, but they just aren't running at all right now!
See e.g. https://github.com/Layr-Labs/eigenlayer-contracts/actions/runs/6525043953/job/17717316889?pr=247#step:5:120
I also think it being split into 77 different runs might be a tad much -- can we split this any more sanely, perhaps into like 5-10 suites?

File: src/contracts/pods/EigenPod.sol [hunk 1]
```diff
@@ -672,7 +673,14 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
...
         validatorInfo.status = VALIDATOR_STATUS.WITHDRAWN;
-        validatorInfo.mostRecentBalanceUpdateTimestamp = withdrawalTimestamp;
+        /**
+        * if multiple full withdrawals are proven for the same validator, we ensure the latest timestamp is recorded
+        * if the withdrawals are proven out of order. A validator may have to make multiple full withdrawals if 
+        * they redeposit into their exited validator.
+        */
+        if (validatorInfo.mostRecentBalanceUpdateTimestamp < withdrawalTimestamp){
+            validatorInfo.mostRecentBalanceUpdateTimestamp = withdrawalTimestamp;   
+        }
```
Comment (ChaoticWalrus) [inline src/contracts/pods/EigenPod.sol:683]: so this change is specifically in reference to https://github.com/Layr-Labs/eigenlayer-contracts/issues/240 -- right? 
I commented some on the issue in Slack -- seems like we need to agree on the correct behavior before merging this change.

File: src/test/EigenPod.t.sol [hunk 2]
```diff
@@ -1488,8 +1487,8 @@ contract EigenPodTests is ProofParsing, EigenPodPausingConstants {
         address recipient,
         uint64 withdrawalAmountGwei,
         ValidatorInfo memory validatorInfo
-    ) public {
-        _processFullWithdrawal(
+    ) public returns(IEigenPod.VerifiedWithdrawal memory) {
+        return _processFullWithdrawal(
```
Comment (ChaoticWalrus) [inline src/test/EigenPod.t.sol:1491]: why this change? seems sensible + presumably better for integrations, but noting that this changes the interface.

Comment (ChaoticWalrus) [reply src/test/EigenPod.t.sol:1491]: nvm, test file lol

File: src/test/EigenPod.t.sol [hunk 3]
```diff
@@ -1498,4 +1497,18 @@ contract EigenPodTests is ProofParsing, EigenPodPausingConstants {
...
+        address recipient,
+        uint64 withdrawalAmountGwei
+    ) public returns(IEigenPod.VerifiedWithdrawal memory) {
+        return _processPartialWithdrawal(
+            validatorIndex,
+            withdrawalHappenedTimestamp,
+            recipient,
+            withdrawalAmountGwei
+        );
+    }
```
Comment (ChaoticWalrus) [inline src/test/EigenPod.t.sol:1513]: I think this 'wrapper' contract (that you've named `EPInternalFunctions`) should probably be broken out into a separate file

File: src/test/unit/EigenPodUnit.t.sol [hunk 4]
```diff
@@ -275,4 +275,143 @@ contract EigenPodUnitTests is EigenPodTests {
...
+        cheats.deal(address(this), amount);
+        // simulate a withdrawal processed on the beacon chain, pod balance goes to 32 ETH
+        Address.sendValue(payable(address(newPod)), amount);
+        require(newPod.nonBeaconChainETHBalanceWei() == amount, "nonBeaconChainETHBalanceWei should be 32 ETH");
+
+        cheats.startPrank(podOwner);
+        newPod.withdrawBeforeRestaking();
+        cheats.stopPrank();
+
+        require(_getLatestDelayedWithdrawalAmount(podOwner) == amount, "Payment amount should be stake amount");
```
Comment (ChaoticWalrus) [inline src/test/unit/EigenPodUnit.t.sol:317]: can we also check that whatever the name is for ETH sent to the EigenPod via the fallback is now zero?

File: src/test/unit/EigenPodUnit.t.sol [hunk 5]
```diff
@@ -275,4 +275,143 @@ contract EigenPodUnitTests is EigenPodTests {
...
+        });
+        podInternalFunctionTester.processFullWithdrawal(0, pubkeyHash, timestamp += 5 , podOwner, withdrawalAmount, validatorInfo);
+        IEigenPod.ValidatorInfo memory info = podInternalFunctionTester.validatorPubkeyHashToInfo(pubkeyHash);
+
+        podInternalFunctionTester.processFullWithdrawal(0, pubkeyHash, lesserTimestamp , podOwner, withdrawalAmount, podInternalFunctionTester.validatorPubkeyHashToInfo(pubkeyHash));
+
+        IEigenPod.ValidatorInfo memory info2 = podInternalFunctionTester.validatorPubkeyHashToInfo(pubkeyHash);
+
+        require(info2.mostRecentBalanceUpdateTimestamp == timestamp, "mostRecentBalanceUpdateTimestamp should not have been updated");
+    }
```
Comment (ChaoticWalrus) [inline src/test/unit/EigenPodUnit.t.sol:338]: this should have a comment or something at least explaining what the intended behavior is here.
The naming makes it sound like maybe it should revert to me (which it shouldn't)

File: src/test/unit/EigenPodUnit.t.sol [hunk 6]
```diff
@@ -275,4 +275,143 @@ contract EigenPodUnitTests is EigenPodTests {
...
+            status: IEigenPod.VALIDATOR_STATUS.ACTIVE
+        });
+        IEigenPod.VerifiedWithdrawal memory vw = podInternalFunctionTester.processFullWithdrawal(0, pubkeyHash, 0, podOwner, withdrawalAmount, validatorInfo);
+
+        if(withdrawalAmount > podInternalFunctionTester.MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR()){
+            require(vw.amountToSendGwei == withdrawalAmount - podInternalFunctionTester.MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR(), "newAmount should be MAX_RESTAKED_BALANCE_GWEI_PER_VALIDATOR");
+        }
+        else{
+            require(vw.amountToSendGwei == 0, "newAmount should be withdrawalAmount");
+        }
```
Comment (ChaoticWalrus) [inline src/test/unit/EigenPodUnit.t.sol:355]: shouldn't you also check that the 2 amounts add up to the total?

**Comment by Sidu28** [comment]
After discussion with @ChaoticWalrus , we decided to simply not update mostRecentBalanceupdateTimestamp at all in processFullWithdrawal.  It was unnecessary in the first place.  Relevant issue: #240

File: src/contracts/pods/EigenPod.sol (hunk 1 above)
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol]: Resolved!

File: src/test/EigenPod.t.sol (hunk 2 above)
Comment (Sidu28) [reply src/test/EigenPod.t.sol:1491]: haha yeah just to test the VerifiedWithdrawal being correct

File: src/test/EigenPod.t.sol (hunk 3 above)
Comment (Sidu28) [reply src/test/EigenPod.t.sol:1513]: Fair, done

File: src/test/unit/EigenPodUnit.t.sol (hunk 4 above)
Comment (Sidu28) [reply src/test/unit/EigenPodUnit.t.sol:317]: added

File: src/test/unit/EigenPodUnit.t.sol (hunk 5 above)
Comment (Sidu28) [reply src/test/unit/EigenPodUnit.t.sol]: this test is no longer relevant

File: src/test/unit/EigenPodUnit.t.sol (hunk 6 above)
Comment (Sidu28) [reply src/test/unit/EigenPodUnit.t.sol:335]: discussed offline, resolved
